Binding constraint RHS scenarization

JasonMarechal25 opened the pull request:
- [x] Does Parameters need a "refresh" member for BC ?
- [X] genericStoreTimeseriesNumbers has been duplicated. Find a better solution (libs/solver dependency, architecture) (see ticket)
- [x] fixTSNumbersWhenWidthIsOne handle BC? **yes**
- [x] Save time series
- [x] Load time series
- [x] Load TSNumbers (scenario builder)
- [x] Apply TSNumbers to Time series
- [x] Check/handle 1 column for a file. Apply value for all series.
- [x] Check/handle empty file for time series
- [x] Check/handle empty file for TSNumbers if applicable
- [x] loadFromCSVFile: Do we need to handle hourly/weekly/daily differently? (3 different size)
- [x] Handle study < 860 and study > 860
- [x] performWL: What is WL? Do we need to add BC ?
- [x] Can lt/gt/eq files be missing? Smart/check: Only try to load data for corresponding constraints type
- [x] Study updater
- [x] Check RuntimeInfos
- [x] update congestionProbabilitiy for BC ?
- [x] If version <= 870, one single group containing all BCs

JasonMarechal25 committed:
Extract methode

JasonMarechal25 committed:
Remove useless variable

JasonMarechal25 committed:
Break some cyclical include dependencies

JasonMarechal25 committed:
Add new unit test for binding constraints

flomnes commented on src/libs/antares/study/area/area.cpp:
```diff
@@ -36,6 +36,8 @@
 #include "ui.h"
 #include "scratchpad.h"
+#include "antares/study/parts/parts.h"
+#include "antares/study/parts/load/prepro.h"
```
This should be included in parts.h

JasonMarechal25 replied:
It's complicated. Including parts.h and having prepro included by parts.h create some cyclical inclusion.

flomnes reviewed (commented):
LGTM

JasonMarechal25 committed:
BindingConstraint group property: RED

JasonMarechal25 committed:
BindingConstraint group property: GREEN

JasonMarechal25 committed:
Refactor test: generate file in TEMP instead of datafiles

JasonMarechal25 committed:
BoundingConstraints load group: RED

JasonMarechal25 committed:
BoundingConstraints load group: GREEN

JasonMarechal25 committed:
BoundingConstraints load RHS properly

JasonMarechal25 committed:
Store TS Numbers

JasonMarechal25 committed:
Save scenario builder for BC

JasonMarechal25 committed:
Save scenario builder for BC with all rules

JasonMarechal25 committed:
Use matrix type to save time series numbers

JasonMarechal25 committed:
Properly load time series numbers

JasonMarechal25 committed:
Fix store time series numbers

JasonMarechal25 committed:
Add test store timeseries number to unit tests

JasonMarechal25 committed:
Renaming constraint -> bindingConstraint

JasonMarechal25 committed:
Update for BCTimeSeries [FAIL]

JasonMarechal25 committed:
Load BC TimeSeries [PASS]

JasonMarechal25 committed:
Test that reading different TS size fail [PASS]

JasonMarechal25 committed:
Test that all BC have the same type [PASS]

JasonMarechal25 committed:
Rename "series" member of DataSeries types as "time_series"

JasonMarechal25 committed:
Fix compilation for all plateformes

flomnes commented:
`performWL` seems to be part of a library of functions whose intent is to perform a queue of actions.

This library is only used in the (deprecated) GUI for copy-pasting areas or links. It may also be used in other context such as renaming or removing objects, but I'm not sure.

I think it's best to stay away from it.

payetvin commented on src/libs/antares/study/binding_constraint/BindingConstraintTimeSeries.cpp:
```diff
@@ -0,0 +1,22 @@
+
+namespace Antares::Data {
+
+    unsigned BindingConstraintTimeSeries::SeriesWidth() const {
```
we use this convention for braces 
``` 
const
{

payetvin commented on src/libs/antares/study/binding_constraint/BindingConstraintTimeSeries.h:
```diff
@@ -0,0 +1,24 @@
+
+class BindingConstraintTimeSeries {
+public:
+    BindingConstraintTimeSeries() = default;
```
= default is needed here ?

JasonMarechal25 committed:
Simple test binding constraint use time series [FAIL]

JasonMarechal25 committed:
Use vector instead of c-style array for binding constraints RTI

JasonMarechal25 committed:
Fix resizeAllTimeseriesNumbers

JasonMarechal25 committed:
Simple test binding constraint use time series [PASS]

JasonMarechal25 committed:
Rename upper_bound/lower_bound time series as greater_than/lesser_than to avoid confusion

JasonMarechal25 committed:
Simple test binding constraint use time series [PASS]

JasonMarechal25 committed:
Time series are scenarised by BindingConstraint not groups

flomnes commented on src/libs/antares/study/area/area.hxx:
```diff
@@ -74,6 +74,8 @@ inline void Area::storeTimeseriesNumbers(Solver::IResultWriter::Ptr writer) cons
         storeTimeseriesNumbersForTransmissionCapacities(writer);
         break;
+    case timeSeriesBindingConstraints:
+        storeTimeseriesNumbersForBindingConstraints(writer);
```
`break` is not strictly necessary, but might prevent a bug if new `case`s are added in the future.

flomnes commented on src/libs/antares/study/area/list.cpp:
```diff
@@ -855,7 +857,7 @@ static bool AreaListLoadFromFolderSingleArea(Study& study,
     if (study.header.version <= 320)
     {
-        area.solar.series->series.resize(1, HOURS_PER_YEAR);
+        area.solar.series->time_series.resize(1, HOURS_PER_YEAR);
```
This code no longer exists in develop, consider merging develop into this branch

JasonMarechal25 committed:
WIP Time series in BC

JasonMarechal25 committed:
Only load timeseries for the bionding constraint operator

JasonMarechal25 committed:
Use time_series and ts_numbers for simulation

JasonMarechal25 commented on src/libs/antares/study/binding_constraint/BindingConstraint.cpp:
```diff
@@ -731,6 +724,14 @@ bool BindingConstraint::loadFromEnv(BindingConstraint::EnvForLoading& env)
                          << "]: Invalid operator [less,greater,equal,both]";
+        if (group_.empty()) {
+            if (env.version >= version860) {
+                logs.error() << env.iniFilename << ": in [" << env.section->name
```
We should probably return "false" as error

flomnes commented on src/libs/antares/CMakeLists.txt:
```diff
@@ -499,7 +499,7 @@ set(SRC
     Enum.hpp
     Enum.hxx
-)
+        study/scenario-builder/BindingConstraintsTSNumbersData.cpp study/binding_constraint/BindingConstraintTimeSeries.cpp study/binding_constraint/BindingConstraintTimeSeries.h study/binding_constraint/BindingConstraintTimeSeriesNumbers.cpp study/binding_constraint/BindingConstraintTimeSeriesNumbers.h)
```
One file per line for better readability.

flomnes commented on src/libs/antares/study/area/area.cpp:
```diff
@@ -155,8 +150,8 @@ AreaLink* Area::findExistingLinkWith(Area& with)
-        const AreaLink::Map::iterator end = with.links.end();
-        for (AreaLink::Map::iterator i = with.links.begin(); i != end; ++i)
+        const auto end = with.links.end();
+        for (auto i = with.links.begin(); i != end; ++i)
```
```suggestion
        for (auto i = with.links.begin(); i != with.links.end(); ++i)
```

flomnes commented on src/libs/antares/study/area/area.h:
```diff
@@ -38,6 +38,7 @@
 #include <set>
 #include "links.h"
 #include "ui.h"
+#include "antares/study/parameters/adq-patch-params.h"
```
Why is this needed here ?

flomnes commented on src/libs/antares/study/binding_constraint/BindingConstraint.h:
```diff
@@ -241,9 +254,13 @@ class BindingConstraint final : public Yuni::NonCopyable<BindingConstraint>
 
+    //Ref to prevent copy. const ref to prevent modification.
+    const Matrix<>& TimeSeries() const;
+    Matrix<> &TimeSeries();
```
```suggestion
    Matrix<>& TimeSeries();
```

flomnes commented on src/libs/antares/study/binding_constraint/BindingConstraint.h:
```diff
@@ -216,6 +220,15 @@ class BindingConstraint final : public Yuni::NonCopyable<BindingConstraint>
+    ** \brief Get the group
+    */
+    std::string group() const;
+    void group(std::string group_name);
```
```suggestion
    void setGroupName(std::string group_name);
```

flomnes commented on src/libs/antares/study/binding_constraint/BindingConstraintTimeSeriesNumbers.cpp:
```diff
@@ -0,0 +1,11 @@
+
+namespace Antares::Data {
+    unsigned BindingConstraintTimeSeriesNumbers:: SeriesWidth() const {
+        return 0;
```
TODO ?

flomnes commented on src/libs/antares/study/fwd.h:
```diff
@@ -222,6 +222,7 @@ enum TimeSeries
     timeSeriesRenewable = 32,
     //! TimeSeries : Renewable
     timeSeriesTransmissionCapacities = 64,
+    timeSeriesBindingConstraints = 128,
```
timeSeriesCount = 8

flomnes replied:
Used only for TS generation

flomnes commented on src/libs/antares/study/parameters.h:
```diff
@@ -271,6 +271,7 @@ class Parameters final
     uint nbTimeSeriesThermal;
     //! Nb of timeSeries : Solar
     uint nbTimeSeriesSolar;
+    uint nbTimeSeriesBindingConstraints;
```
This is not required. `nbTimeSeriesThermal` and others are only used to generate a certain number of TS by the TS generator.

We do not intend to generate TS for BC RHS.

flomnes commented on src/libs/antares/study/parameters.h:
```diff
@@ -293,6 +294,7 @@ class Parameters final
     uint refreshIntervalThermal;
     //! Refresh interval (in years) for timeSeries : Solar
     uint refreshIntervalSolar;
+    //TODO BC ?
```
Same as above, refresh does not apply here since we don't intend to generate TS for BC RHS.

flomnes commented on src/libs/antares/study/parts/load/container.h:
```diff
@@ -37,6 +36,8 @@ namespace Data
 namespace Load
 {
+
+class Prepro;
```
What's the deal with load prepro ? How is this related to BCs ?

flomnes commented on src/libs/antares/study/runtime/runtime.cpp:
```diff
@@ -524,15 +527,15 @@ static bool isBindingConstraintTypeInequality(const Data::BindingConstraintRTI&
-    const auto* firstBC = this->bindingConstraint;
-    const auto* lastBC = firstBC + this->bindingConstraintCount;
+    const auto firstBC = bindingConstraint.begin();
+    const auto lastBC = bindingConstraint.end();
```
This piece of code was removed in branch develop, you might get conflicts at merge

flomnes commented on src/libs/antares/study/scenario-builder/TSnumberData.h:
```diff
@@ -314,8 +310,49 @@ inline CString<512, false> ntcTSNumberData::get_prefix() const
+    return "bc,";
+}
+
+inline unsigned BindingConstraintsTSNumberData::get_tsGenCount(const Study& study) const {
+    auto& parameters = study.parameters;
+
+    bool genTS = (0 != (parameters.timeSeriesToGenerate & timeSeriesBindingConstraints));
+    return genTS ? 1 : 0u;
+}
```
`timeSeriesToGenerate & timeSeriesBindingConstraints` should always be 0, no TS is generated for BCs

flomnes commented on src/solver/simulation/CMakeLists.txt:
```diff
@@ -40,7 +40,7 @@ set(SRC_SIMULATION
 		adequacy_patch_runtime_data.h
 		adequacy_patch_runtime_data.cpp
-		)
+		ITimeSeriesWriter.h TimeSeriesWriter.cpp TimeSeriesWriter.h)
```
One file per line

flomnes commented on src/solver/simulation/TimeSeriesWriter.h:
```diff
@@ -0,0 +1,23 @@
+class TimeSeriesWriter: public ITimeSeriesWriter {
+
+public:
+    TimeSeriesWriter(std::shared_ptr<Antares::Solver::IResultWriter> sharedPtr);
```
```suggestion
    TimeSeriesWriter(std::shared_ptr<Antares::Solver::IResultWriter> writer);
```

flomnes commented on src/solver/simulation/sim_calcul_economique.cpp:
```diff
@@ -358,28 +359,31 @@ void SIM_RenseignementProblemeHebdo(PROBLEME_HEBDO& problem,
                 for (uint d = 0; d != 7; ++d)
                 {
-                    sndMember[d] = column[weekFirstDay + d];
+                    //TODO
+                    //sndMember[d] = column[weekFirstDay + d];
```
This is the "core" of the development, where all the magic takes place !

flomnes commented on src/solver/simulation/timeseries-numbers.cpp:
```diff
@@ -949,20 +952,21 @@ bool TimeSeriesNumbers::Generate(Study& study)
-        study.storeTimeSeriesNumbers<timeSeriesThermal>();
-        study.storeTimeSeriesNumbers<timeSeriesRenewable>();
-        study.storeTimeSeriesNumbers<timeSeriesTransmissionCapacities>();
+        study.storeTimeSeriesNumbers<TimeSeries::timeSeriesLoad>();
+        study.storeTimeSeriesNumbers<TimeSeries::timeSeriesSolar>();
+        study.storeTimeSeriesNumbers<TimeSeries::timeSeriesHydro>();
+        study.storeTimeSeriesNumbers<TimeSeries::timeSeriesWind>();
+        study.storeTimeSeriesNumbers<TimeSeries::timeSeriesThermal>();
+        study.storeTimeSeriesNumbers<TimeSeries::timeSeriesRenewable>();
+        study.storeTimeSeriesNumbers<TimeSeries::timeSeriesTransmissionCapacities>();
```
Is it planned to use `writer` here as well ?

guilpier-code replied:
> Is it planned to use `writer` here as well ?

I think we're forced to do it like that for now : the above calls to `study.storeTimeSeriesNumbers<T>()` internally loop over areas and use the same writer internally as well.
But we can't extend these **storeTimeSeriesNumbers\<T\>()** for BCs, as they are not related to areas.

flomnes commented on src/solver/aleatoire/alea_tirage_au_sort_chroniques.cpp:
```diff
@@ -185,6 +187,23 @@ static void InitializeTimeSeriesNumbers_And_ThermalClusterProductionCost(
         ptchro.TransmissionCapacities
           = (directWidth != 1) ? link->timeseriesNumbers[0][year] : 0; // zero-based
     }
+    //Each binding constraints group
+    std::vector<std::string> all_groups;
+    std::transform(runtime.bindingConstraint.begin(), runtime.bindingConstraint.end(),
+                   std::back_inserter(all_groups),
+                   [](auto& bc)->std::string { //Catch by reference else risk double delete of inner pointer members
+        return bc.group;
+    });
+    all_groups.erase(std::unique(all_groups.begin(), all_groups.end()), all_groups.end());
+    for (const auto& group_name : all_groups) {
+        for (const auto& bc: runtime.bindingConstraint)
+        {
+            if (bc.group == group_name) {
+                assert(year < bc.time_series.width); //TODO width or height ?
+            }
+        }
+        NumeroChroniquesTireesParGroup[numSpace][group_name] = study.bindingConstraints.time_series[group_name].timeseriesNumbers[0][year];
```
You do 2 things here
- Loop on groups
- In debug mode, check consistency `year` / `bc.time_series.width`

In my opinion, it's not necessary to construct a list of groups here, you could simply loop on BCs

```suggestion
    //Each binding constraints group
    std::vector<std::string> all_groups;
    std::transform(runtime.bindingConstraint.begin(), runtime.bindingConstraint.end(),
                   std::back_inserter(all_groups),
                   [](auto& bc)->std::string { //Catch by reference else risk double delete of inner pointer members
        return bc.group;
    });
    all_groups.erase(std::unique(all_groups.begin(), all_groups.end()), all_groups.end());
    for (const auto& group_name : all_groups) {
        for (const auto& bc: runtime.bindingConstraint)
        {
            if (bc.group == group_name) {
                assert(year < bc.time_series.width); //TODO width or height ?
            }
        }
        NumeroChroniquesTireesParGroup[numSpace][group_name] = study.bindingConstraints.timeseriesNumbers(group_name, year);
```

flomnes reviewed (changes requested):
I haven't reviewed tests, will do at next review

You can remove `Parameters::nbTimeSeriesBindingConstraints` and everything related

JasonMarechal25 committed:
Handle Daily BC

JasonMarechal25 committed:
Binding constraints weekly

guilpier-code commented on src/libs/antares/study/runtime/runtime.h:
```diff
@@ -98,6 +88,8 @@ class BindingConstraintRTI
     long* clustersAreaIndex;
     std::string name;
+    std::string group;
+    Matrix<> time_series;
```
**time_series** ==> **rhs_time_series** ? (The information that this regards RHS is missing)

JasonMarechal25 committed:
Refactor test

JasonMarechal25 committed:
Lesser and Greater inequality binding constraints

guilpier-code commented on src/libs/antares/study/binding_constraint/BindingConstraint.h:
```diff
@@ -511,6 +533,10 @@ class BindingConstraint final : public Yuni::NonCopyable<BindingConstraint>
     //! Comments
     YString pComments;
+    //! Group
+    std::string group_;
```
**Renaming** :
As it is a string, **group_** ==> **group_name_** ?

guilpier-code commented on src/libs/antares/study/binding_constraint/BindingConstraint.h:
```diff
@@ -131,6 +133,8 @@ class BindingConstraint final : public Yuni::NonCopyable<BindingConstraint>
         //! List of areas
         AreaList& areaList;
+
+        unsigned version;
```
**Renaming**
version ==> **study_version** ?

guilpier-code commented on src/libs/antares/study/binding_constraint/BindingConstraint.h:
```diff
@@ -482,13 +499,18 @@ class BindingConstraint final : public Yuni::NonCopyable<BindingConstraint>
-    Matrix<> pValues;
+    //Matrix<> pValues;
+    //! Time series of the binding constraint. Width = number of series. Height = nbTimeSteps. Only store series for operatorType
+    Matrix<> time_series;
```
Renaming : 
time_series ==> **rhs_time_series_** (private attribute : trailing "_")

guilpier-code commented on src/libs/antares/study/binding_constraint/BindingConstraint.cpp:
```diff
@@ -268,8 +269,10 @@ void BindingConstraint::resetToDefaultValues()
+    time_series.reset();
+    markAsModified();
+    //pValues.zero();
+    //pValues.markAsModified();
```
To remove : 
```C++
//pValues.zero();
//pValues.markAsModified();
```

guilpier-code commented on src/libs/antares/study/binding_constraint/BindingConstraint.cpp:
```diff
@@ -716,7 +709,7 @@ bool BindingConstraint::loadFromEnv(BindingConstraint::EnvForLoading& env)
 
     // Checking for validity
-    if (!ret or !pName or !pID or pOperator == opUnknown or pType == typeUnknown)
+    if (!pName or !pID or pOperator == opUnknown or pType == typeUnknown)
```
About the **if** statement : 
```cpp
if (!pName or !pID or pOperator == opUnknown or pType == typeUnknown)
```
is it useful ? (as we test every data member inside the **if** anyway).
If it's useful, we should add a new test : `!group_.empty()`

guilpier-code commented on src/libs/antares/study/binding_constraint/BindingConstraint.cpp:
```diff
@@ -1537,10 +1530,75 @@ BindingConstraint* BindingConstraintsList::add(const AnyString& name)
+}
+
+bool
+BindingConstraint::loadBoundedTimeSeries(BindingConstraint::EnvForLoading &env, BindingConstraint::Operator operatorType) {
```
Renaming :  
**loadBoundedTimeSeries(...)** : we gave up the **bound** thing in the TS attribute name (**time_series**).
Should be renamed into **loadRHStimeSeries(...)**

Besides shouldn't this method be **private** ?

guilpier-code commented on src/libs/antares/study/binding_constraint/BindingConstraint.cpp:
```diff
@@ -1537,10 +1530,75 @@ BindingConstraint* BindingConstraintsList::add(const AnyString& name)
+        logs.info() << " loaded time series for `" << name() << "` (" << BindingConstraint::TypeToCString(type()) << ", "
+                    << BindingConstraint::OperatorToShortCString(operatorType) << ')';
+        return true;
+    } else {
```
No logs in case we fail loading ?

guilpier-code commented on src/libs/antares/study/binding_constraint/BindingConstraint.cpp:
```diff
@@ -506,24 +509,9 @@ void BindingConstraint::copyOffsets(const Study& study,
 }
 
-bool BindingConstraint::loadFromEnv(BindingConstraint::EnvForLoading& env)
+bool BindingConstraint::loadFromEnv(BindingConstraint::EnvForLoading& env, unsigned nb_years)
```
**loadFromEnv(...)** : parameter **nb_years** seems useless

JasonMarechal25 committed:
Refactor: squash a switch case split in two

guilpier-code commented on src/libs/antares/study/binding_constraint/BindingConstraint.cpp:
```diff
@@ -1537,10 +1530,75 @@ BindingConstraint* BindingConstraintsList::add(const AnyString& name)
     return bc;
 }
 
+void BindingConstraintsList::resizeAllTimeseriesNumbers(unsigned int nb_years) {
```
**resizeAllTimeseriesNumbers** : I can't see any place where this function is called. But maybe I missed something.

guilpier-code replied:
**Question** : 
there are 2 loops here, and I'm confused : 
We fill the `std::map` **time_series** (which ill named by the way, see comment in the header file).
At this stage isn't this map supposed to already contain the keys, which are precisely the group names ?
In that case, do we need the first loop ?

guilpier-code commented on src/libs/antares/study/binding_constraint/BindingConstraint.h:
```diff
@@ -666,13 +692,21 @@ class BindingConstraintsList final : public Yuni::NonCopyable<BindingConstraints
     BindingConstraint::Vector pList;
 
+public:
+    std::map<std::string, BindingConstraintTimeSeriesNumbers> time_series;
```
Renaming : 
**time_series** : should be rename into **groupToTSnumbers** (for example)

guilpier-code commented on src/libs/antares/study/area/area.hxx:
```diff
@@ -24,14 +24,14 @@
-#define __ANTARES_LIBS_STUDY_AREAS_HXX__
+#pragma once
 
+#include "area.h"
```
`#include "area.h"` : 
the current **area.hxx** is included at the end of "area.h", why do we need to include **area.h** in **area.hxx** back ?

guilpier-code replied:
Same question for **parts.h**

JasonMarechal25 commented:
>  fixTSNumbersWhenWidthIsOne handle BC? **yes**

This is more complicated than it seems. This function set the TSNumber to "one" for each year when there is only one time series. The effect is that we select the first time series for each year, which is the intent since their is only one TS.
For binding constraints it's different. In a group we may have constraints with only one TS mixed with constraints with N TS.
We can update fixTSNumbersWhenWidthIsOne for the case where all TS in a group have width=1 but we also need a way to get the first (and only) TS when asking for the N-th TS of a constraint that only have one.

JasonMarechal25 committed:
Rename time_series -> time_series_numbers for structure in BindingConstraintsList

JasonMarechal25 committed:
Binding Constraints works with multiple time series

JasonMarechal25 committed:
Binding Constraints works with several years only one TS

JasonMarechal25 committed:
Binding Constraints works with several years only one TS

JasonMarechal25 committed:
Load time_series empty files as a column of 0s

JasonMarechal25 committed:
Fix for 1 TS several years

JasonMarechal25 committed:
Empty TS file is an error (BC not loaded)

payetvin committed:
[DEV] Add legacy loading function for time series

payetvin committed:
[GIT] Modify githook regex to allow capitals

JasonMarechal25 committed:
RNG ts numbers for BC

payetvin committed:
[FIX] Add intermediate matrix

JasonMarechal25 committed:
Handle range type

JasonMarechal25 committed:
Handle range type

JasonMarechal25 committed:
Use shared_ptr instead of raw pointers for binding constraints

JasonMarechal25 committed:
f

JasonMarechal25 committed:
Code clean up

JasonMarechal25 committed:
Extract methode "SeparateValue"

Remove duplication to read Link or cluster values while parsing binding constraints

JasonMarechal25 committed:
Fix ui compilation

JasonMarechal25 committed:
Rename to StoreTimeSeriesNumbersIntoOuput to be accurate

JasonMarechal25 committed:
Remove unimplemented function to store BC TSNumbers

JasonMarechal25 committed:
Initialize TSNumbers so that they exist for sizing and filling

JasonMarechal25 committed:
Properly set TSNumbers from RNG.

JasonMarechal25 committed:
Assert when trying to load wrong operator BC for study pre-860

JasonMarechal25 committed:
Remove dead code

JasonMarechal25 committed:
Extract code to save bBindingCosntraint in dedicated class

JasonMarechal25 committed:
Constness and remove "not" for "!"

JasonMarechal25 committed:
Refactoring: minor sonar issues

JasonMarechal25 committed:
Move time series loading into BindingConstraintLoader

JasonMarechal25 committed:
Refactor building file_name to save/load time series

JasonMarechal25 committed:
Code cleanup: DeadCode, Sonar issues, etc.

JasonMarechal25 committed:
Update memoryUsage

JasonMarechal25 committed:
Remove dead code BindingConstraintTimeSeries

JasonMarechal25 committed:
Minor fix

JasonMarechal25 committed:
Remove nbTimeSeriesBindingConstraints in parameters

JasonMarechal25 committed:
Fix post rebase

JasonMarechal25 committed:
Remove TODO

JasonMarechal25 committed:
Binding constraints are never used with TS-generator

JasonMarechal25 committed:
time_series -> TimeSeries

JasonMarechal25 committed:
Fix warnings

JasonMarechal25 committed:
Remove return value of "ForceReload"

This value is never used

JasonMarechal25 committed:
Update constraintBuilder

JasonMarechal25 committed:
Update ensureDataAreLoadedForAllBindingConstraints

JasonMarechal25 committed:
Check all time series in a group have the same width

flomnes commented on src/libs/antares/study/binding_constraint/BindingConstraintsList.cpp:
```diff
@@ -0,0 +1,387 @@
+    std::for_each(pList.begin(), pList.end(), [&groupOfOneTS](auto bc) {
+        auto hasOneTs = bc->TimeSeries().width == 1;
+        if (groupOfOneTS[bc->group()] && !hasOneTs) {
+            assert(false && ("Group of binding constraints mixing 1TS and N TS group:" + bc->group()).c_str());
```
Assert is only executed in debug mode. You should issue an error instead, see `logs.error()`

flomnes commented on src/libs/antares/study/binding_constraint/BindingConstraintsList.cpp:
```diff
@@ -0,0 +1,387 @@
+        std::map<std::string, unsigned, std::less<>> timeSeriesCountByGroup;
+        for(const auto& bc: this->pList) {
+            auto count = timeSeriesCountByGroup[bc->group()];
+            auto width = bc->TimeSeries().width;
+            if (count == 0) {
```
Some BCs may have a single time-series, this input should be considered valid.
```suggestion
            auto width = bc->TimeSeries().width;
            if (width == 1) continue;
            if (count == 0) {
```

flomnes commented on src/libs/antares/study/binding_constraint/BindingConstraintsList.cpp:
```diff
@@ -0,0 +1,387 @@
+                             << bc->group()
+                             << " .Previous width was " << count
+                             << " new constraint " << bc->name()
+                             << "found with width of " << width;
```
```suggestion
                             << " found with width of " << width;
```

flomnes commented on src/libs/antares/study/binding_constraint/BindingConstraintsList.cpp:
```diff
@@ -0,0 +1,387 @@
+        return hasError;
+    }
+
+    void BindingConstraintsList::InitializeTSNumbers() {
```
```suggestion
    void BindingConstraintsList::initializeTSNumbers() {
```

flomnes commented on src/libs/antares/study/binding_constraint/BindingConstraintsList.cpp:
```diff
@@ -0,0 +1,387 @@
+    }
+
+    void BindingConstraintsList::InitializeTSNumbers() {
+    std::set<std::string> groups;
+    std::for_each(this->pList.begin(), this->pList.end(), [&groups](auto bc) {
+        groups.insert(bc->group());
+    });
+    for (const auto& group: groups) {
+        this->timeSeriesNumbers[group] = {};
+    }
```
In this short function, it's quite obvious that `pList` is a data member, so you could just drop the `this->`.

Also, this temporary `std::set` probably isn't necessary. You could do this in a single loop if you accept that you assign `timeSeriesNumbers[group] = {}` multiple times.

JasonMarechal25 committed:
Simplify InitializeTSNumbers

JasonMarechal25 committed:
Simplification

Co-authored-by: Florian Omnès <florian.omnes@rte-france.com>

JasonMarechal25 committed:
f time series count

JasonMarechal25 committed:
f time series count

JasonMarechal25 committed:
Remove dead code

JasonMarechal25 committed:
Reintroduce code wrongly deleted in constraintBuilder

JasonMarechal25 committed:
Minor fix

Co-authored-by: Florian Omnès <florian.omnes@rte-france.com>

JasonMarechal25 committed:
Fix virtual issues in TimeSeriesNumbersWriter

JasonMarechal25 committed:
Add log in case loadTimeSeries fail

JasonMarechal25 committed:
Fix compilation error

JasonMarechal25 committed:
Remove bindingConstraint value from timeSeries enum

Only used for ts generator and intramodal.

JasonMarechal committed:
Fix windows compilation issue

guilpier-code commented on src/solver/simulation/timeseries-numbers.cpp:
```diff
@@ -950,20 +961,21 @@ bool TimeSeriesNumbers::Generate(Study& study)
 
     if (study.parameters.storeTimeseriesNumbers)
     {
         fixTSNumbersWhenWidthIsOne(study);
```
**Does not strictly concern this PR**
To me, instruction `fixTSNumbersWhenWidthIsOne(study);` has nothing to do here, as it is not strictly related to TS storing into output, but changes the time series themselves.
So I guess It should be called at the end of **TimeSeriesNumbers::Generate(...)**.

guilpier-code commented on src/solver/simulation/solver.hxx:
```diff
@@ -379,7 +380,8 @@ void ISimulation<Impl>::run()
 
         // Export ts-numbers into output
-        TimeSeriesNumbers::StoreTimeseriesIntoOuput(study);
+        TimeSeriesNumbersWriter time_seris_writer(pResultWriter);
```
Apparently the class **TimeSeriesNumbersWriter** is related to BC number only.
So why not name it **BindingConstraintTSwriter** (for instance) instead ?

And object **time_seris_writer** should be renamed **ts_bc_writer** (for instance)

Besides, why is ths TS BC writer is instantiated all alone, whereas there are TS from other items that need to be written ?
Shouldn't the BC writer be instantiated inside function **StoreTimeSeriesNumbersIntoOuput** ?

guilpier-code replied:
**Does not strictly apply to this PR** :
And to me, we have a bad design here, the 2 following tasks  : 
- Computing TS number (by function **Generate(...)**)
- Writing the TS numbers (by **StoreTimeSeriesNumbersIntoOuput(...)**)

Should be done out of the **Simulation::run()** and even out of the **Simulation** itself (the first task before, the second after).

guilpier-code commented on src/libs/antares/study/binding_constraint/BindingConstraint.h:
```diff
@@ -216,6 +177,15 @@ class BindingConstraint final : public Yuni::NonCopyable<BindingConstraint>
+    /*!
+    ** \brief Get the group
+    */
+    std::string group() const;
```
`std::string group() const;` to be renamed into `std::string groupName() const;` 
(I was confused when I first met a call to this method : is group() an id ? Is it an object ? Or a name ?)

guilpier-code replied:
Not necessary

guilpier-code commented on src/libs/antares/study/binding_constraint/BindingConstraintsList.cpp:
```diff
@@ -0,0 +1,383 @@
+    std::map<std::string, bool, std::less<>> groupOfOneTS;
+    std::for_each(pList.begin(), pList.end(), [&groupOfOneTS](auto bc) {
+        auto hasOneTs = bc->TimeSeries().width == 1;
+        if (groupOfOneTS[bc->group()] && !hasOneTs) {
```
**Question** :
```C++
if (groupOfOneTS[bc->group()] && !hasOneTs) {...}
```
What happens if the map **groupOfOneTS** does not have the key  **bc->group()** yet ?
Is `groupOfOneTS[bc->group()]` true or false ?

flomnes replied:
@JasonMarechal25 replied alreay

guilpier-code commented on src/libs/antares/study/binding_constraint/BindingConstraintsList.hxx:
```diff
@@ -0,0 +1,75 @@
+}
+
+template<class ListBindingConstraints>
+unsigned int BindingConstraintsList::NumberOfTimeseries(const ListBindingConstraints &list, const std::string& group_name) {
```
**NumberOfTimeseries(...)** : does it need to be a template function ?

flomnes replied:
It seems to be called only once (solver/simulation/timeseries-numbers.cpp:716)

Remove `BindingConstraintRTI` ?

guilpier-code commented on src/libs/antares/study/action/handler/antares-study/constraint/data.cpp:
```diff
@@ -59,15 +59,11 @@ bool Data::performWL(Context& ctx)
-            assert(source->matrix().height > 0);
-            ctx.constraint->matrix() = source->matrix();
-            source->matrix().unloadFromMemory();
+            //Deleted some code. UI is deprecated but not yet removed
```
**Question** : this code is used when requesting a **save as** operation in the historical GUI.
Why do we remove it ?
You removed it because it would have been an extra work to adapt it to the changes you're doing ?

flomnes replied:
This is a broader question : do we keep the GUI ? Or remove it ?

guilpier-code commented on src/libs/antares/study/runtime/runtime.h:
```diff
@@ -163,9 +150,7 @@ class StudyRuntimeInfos
-    //! Number of binding constraint
-    uint bindingConstraintCount;
-    BindingConstraintRTI* bindingConstraint;
+    std::vector<BindingConstraintRTI> bindingConstraint;
```
Rename : 
```C++
std::vector<BindingConstraintRTI> bindingConstraint;
```
into : 
```C++
std::vector<BindingConstraintRTI> RTIbindingConstraintList;
```
?

guilpier-code replied:
Runtime binding constraints are removed in **develop**

guilpier-code commented on src/solver/simulation/timeseries-numbers.cpp:
```diff
@@ -707,6 +711,13 @@ void drawAndStoreTSnumbersForNOTintraModal(const array<bool, timeSeriesCount>& i
             }
         }
     });
+    // Binding constraints
```
**Temporary remark** (can be removed eventually) : 

Here I'm wondering if a class **BindingConstraintGroup** wouldn't be useful.
The idea would be to encapsulate things in a way that makes the code simpler in other code areas.
```C++
class BindingConstraintGroup
{
public:
    // Whatever public function needs to be defined, for instance : 
    BindingConstraintGroup(std::string groupName);
    void addBindingConstraint(BindingConstraint* bc);
    int getTSnumber(int year);
    void setTSnumber(int year);
    bool containedBC(BindingConstraint* bc);
    ...
private:
    std::vector<BindingConstraint*> BCsOfTheGroup;  // Could be a collection of RTI BCs or original BCs 
    vector<int> timeseriesNumbersPerYear; 
    std::string name_;
};

// And maybe : 
class BindingConstraintsGroupList
{
public: 
    // Here some member functions like begin() and end() to facilitate the group collection traversal  
    void addGroup(BindingConstraintGroup* group);
    int getBCtimeseriesNumber(BindingConstraint* bc);
    ...
private:
    std::vector<BindingConstraintGroup*> BCgroups_;
    std::map<BindingConstraint*, BindingConstraintGroup*> bcToBcGroup_;
}
```
As a consequence in **study.h** :  
```C++
class Study
{
    ...
    BindingConstraintsList bindingConstraints; // As usual
    BindingConstraintsGroupList BCgroups;
}
```

And here (in the current source file) :
```C++
for (auto& bcGroup: study.BCgroups)
{
    bcGroup.setTSnumber(year);
}
``` 
What do you think ?

guilpier-code replied:
See issue #1321 to handle this point.

guilpier-code commented on src/libs/antares/study/binding_constraint/BindingConstraint.h:
```diff
@@ -467,28 +392,26 @@ class BindingConstraint final : public Yuni::NonCopyable<BindingConstraint>
-    //! Matrix<> where values for inequalities could be found
-    Matrix<> pValues;
+    //! Time series of the binding constraint. Width = number of series. Height = nbTimeSteps. Only store series for operatorType
+    Matrix<> timeSeries;
```
About `Matrix<> timeSeries;` : 
Should be renamed **rhsTimeSeries**

guilpier-code commented on src/libs/antares/study/binding_constraint/BindingConstraint.h:
```diff
@@ -235,14 +205,10 @@ class BindingConstraint final : public Yuni::NonCopyable<BindingConstraint>
-    Matrix<>& matrix();
+    //Ref to prevent copy. const ref to prevent modification.
+    const Matrix<>& TimeSeries() const;
+    Matrix<>& TimeSeries();
```
`Matrix<>& TimeSeries();`  : 
Should be renamed into **rhsTimeSeries()**

guilpier-code commented on src/libs/antares/study/binding_constraint/BindingConstraintsList.h:
```diff
@@ -0,0 +1,200 @@
+    }
+    void resizeAllTimeseriesNumbers(unsigned nb_years);
+
+    std::map<std::string, Data::BindingConstraintTimeSeriesNumbers, std::less<>> timeSeriesNumbers;
```
For clarity, we should not mix data members and member functions.
We should reject the public data members in a separate **public**  block

guilpier-code replied:
Besides, this data member would not be necessary with the notion of **BindingConstraintGroup** (see remarks in **src/solver/simulation/timeseries-numbers.cpp**)

guilpier-code commented on src/libs/antares/study/binding_constraint/BindingConstraintTimeSeriesNumbers.h:
```diff
@@ -0,0 +1,20 @@
+
+namespace Antares::Data {
+
+class BindingConstraintTimeSeriesNumbers {
```
This class is very small. Therefore, we can question its utility.

Besides, it would not be necessary with the notion of **BindingConstraintGroup** (see remarks in **src/solver/simulation/timeseries-numbers.cpp**)

guilpier-code replied:
Furthermore, **memoryUsage()** is not needed, as we plan to remove the ability to compute an estimated amount of RAM a simulation takes, and an the room the results take on disk

Resolved in [PR 1466](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/1466)

guilpier-code commented on src/solver/simulation/sim_calcul_economique.cpp:
```diff
@@ -327,6 +328,76 @@ void SIM_InitialisationResultats()
+void preparerBindingConstraint(const PROBLEME_HEBDO &problem, uint numSpace, int PasDeTempsDebut,
+                               const StudyRuntimeInfos &studyruntime, const uint weekFirstDay, int pasDeTemps) {
+    const auto constraintCount = studyruntime.bindingConstraint.size();
+    for (uint k = 0; k != constraintCount; ++k)
```
We should rename index **k** into something more meaningful.

guilpier-code commented on src/solver/simulation/sim_calcul_economique.cpp:
```diff
@@ -327,6 +328,76 @@ void SIM_InitialisationResultats()
+        {
+            case BindingConstraint::typeHourly:
+            {
+                double const* column = timeSeries[ts_number];
```
instruction `double const* column = timeSeries[ts_number];` is duplicated.

guilpier-code commented on src/solver/simulation/sim_calcul_economique.cpp:
```diff
@@ -327,6 +328,76 @@ void SIM_InitialisationResultats()
+        assert(bc.timeSeries.width && "Invalid constraint data width");
+        //If there is only one TS, always select it.
+        const auto ts_number = bc.timeSeries.width == 1 ? 0 : NumeroChroniquesTireesParGroup[numSpace][bc.group];
+        auto& timeSeries = bc.timeSeries;
```
`auto& timeSeries = bc.timeSeries;` : not sure it's necessary. **bc.timeSeries** is short enough and more meaningful. Whatever.

guilpier-code commented on src/solver/aleatoire/alea_tirage_au_sort_chroniques.cpp:
```diff
@@ -185,6 +187,14 @@ static void InitializeTimeSeriesNumbers_And_ThermalClusterProductionCost(
+    //Binding constraints
+    //Setting 0 for time_series of width 0 is done when using the value.
+    //To do this here we would have to check every BC for its width
+    for (const auto& [group_name, _] : study.bindingConstraints.timeSeriesNumbers) {
```
According to remarks on **src/solver/simulation/timeseries-numbers.cpp**, we could have : 
```C++
for (group : study.BCgroups)
    NumeroChroniquesTireesParGroup[numSpace][group.name()] = group.getTSnumber(year);
```

guilpier-code replied:
See #1321 to handle this point

guilpier-code commented on src/libs/antares/study/runtime/runtime.cpp:
```diff
@@ -387,7 +369,8 @@ void StudyRuntimeInfos::initializeBindingConstraints(BindingConstraintsList& lis
             rti.linkCount = 0;
-            rti.bounds.clear();
+            rti.timeSeries.clear();
+            //rti.bounds.clear();
```
Remove comment : 
```C++
//rti.bounds.clear();
```

guilpier-code commented on src/libs/antares/study/binding_constraint/BindingConstraintTimeSeriesNumbers.h:
```diff
@@ -0,0 +1,20 @@
+public:
+    Matrix<Yuni::uint32> timeseriesNumbers;
+
+    Yuni::uint64 memoryUsage() const;
```
Not sure we're interested in memory usage anymore. To be discussed.

guilpier-code commented on src/libs/antares/study/scenario-builder/TSnumberData.h:
```diff
@@ -314,8 +310,46 @@ inline CString<512, false> ntcTSNumberData::get_prefix() const
-} // namespace ScenarioBuilder
-} // namespace Data
-} // namespace Antares
+// =================================
```
Please adapt the comment : 
```C++
// =================================
// Transmission capacities ...
// =================================
```
for binding constraints, or binding constraint groups (if we decide to use this notion).

JasonMarechal25 committed:
Class Renaming TimeSeriesNumbersWriter -> BindingConstraintsTimeSeriesNumbersWriter

JasonMarechal25 committed:
Fix typo

JasonMarechal25 committed:
Pluralize vector variable name

JasonMarechal25 committed:
Separate members and functions

JasonMarechal25 committed:
Rename variables for clarity

JasonMarechal25 committed:
Remove duplication

JasonMarechal25 committed:
Remove commented code

JasonMarechal25 committed:
Rename section

guilpier-code commented on src/libs/antares/study/scenario-builder/TSnumberData.h:
```diff
@@ -314,8 +310,46 @@ inline CString<512, false> ntcTSNumberData::get_prefix() const
+    CString<512, false> get_prefix() const override;
+    unsigned get_tsGenCount(const Study& study) const override;
+private:
+    std::map<std::string, MatrixType> rules_;
```
About : 
```C++
std::map<std::string, MatrixType> rules_;
```
If we name the class **BindingConstraintGroupsTSNumberData** (considering notion of BC groups), this private data member becomes of type **MatrixType**, and this data is already in the parent class : 
```C++
protected:
    MatrixType pTSNumberRules;
``` 
It would look more like it's already done for ts number rules of other objets (NTC, hydro, ....). 

Further more,  in `std::map<std::string, MatrixType> rules_;`, do we need a **MatrixType** as value type ? Wouldn't a `std::map<std::string, sd::vector<int>> rules_;` be sufficient ? (As for a group, we need a ts number for each year).

flomnes replied:
Because of function `ApplyToMatrix`

guilpier-code commented on src/libs/antares/study/scenario-builder/rules.cpp:
```diff
@@ -331,6 +333,8 @@ bool Rules::readLine(const AreaName::Vector& splitKey, String value, bool update
     else if (kind_of_scenario == "ntc")
         return readLink(splitKey, value, updaterMode);
+    else if (kind_of_scenario == "bc")
+        return readBindingConstraints(splitKey, value);
```
What about parameter  **updaterMode** ? If it's useless, then it should be removed for everyone, otherwise it should be used here, right ?

guilpier-code replied:
**updaterMode** is useful when modifying the study from old GUI.
Just before modifying a study, for example removing an area from the GUI,  the scenario builder **ini** file is saved as is on disk. 
Right after the area is removed, **updaterMode** is turned to **true**, and the scenario builder previously saved is now loaded, the removed and missing area is ignored by the load, but warnings informing that the area was removed are **NOT** printed. And that's it ! **updaterMode** is used only for that : allowing or not allowing warnings to be printed.
As binding constraints RHS scenarii are not managed by old GUI, we don't care about this **updaterMode** for BC, and there is nothing to do here.

guilpier-code commented on src/libs/antares/study/scenario-builder/BindingConstraintsTSNumbersData.cpp:
```diff
@@ -0,0 +1,75 @@
+
+namespace Antares::Data::ScenarioBuilder
+{
+bool BindingConstraintsTSNumberData::reset(const Study& study)
```
Using the notion of BC groups (see review of **src/solver/simulation/timeseries-numbers.cpp**) and the parent class **TSNumberData**'s `pTSNumberRules` data member to record ts numbers would lead to a simpler function here.
```C++
bool BindingConstraintsTSNumberData::reset(const Study& study)
{
    const uint nbYears = study.parameters.nbYears;
    pTSNumberRules.reset(study.BCgroups.size(), nbYears);
    return true;
}
```

guilpier-code commented on src/libs/antares/study/scenario-builder/BindingConstraintsTSNumbersData.cpp:
```diff
@@ -0,0 +1,75 @@
+**
+** SPDX-License-Identifier: licenceRef-GPL3_WITH_RTE-Exceptions
+*/
+
```
As we have a new source file for BC data definitions, shouldn't we split **src/libs/antares/study/scenario-builder/TSnumberData.cpp** in the same way for load, wind, thermal clusters, etc.

guilpier-code replied:
See issue #1333 to handle this point.

guilpier-code commented on src/libs/antares/study/scenario-builder/rules.h:
```diff
@@ -133,6 +135,8 @@ class Rules final : private Yuni::NonCopyable<Rules>
     bool readSolar(const AreaName::Vector& instrs, String value, bool updaterMode);
     bool readHydroLevels(const AreaName::Vector& instrs, String value, bool updaterMode);
     bool readLink(const AreaName::Vector& instrs, String value, bool updaterMode);
+    bool readBindingConstraints(const AreaName::Vector &splitKey, String value);
```
`bool updaterMode` ?

guilpier-code replied:
**updaterMode** is useful when modifying the study from old GUI.
Just before modifying a study, for example removing an area from the GUI,  the scenario builder **ini** file is saved as is on disk. 
Right after the area is removed, **updaterMode** is turned to **true**, and the scenario builder previously saved is now loaded, the removed and missing area is ignored by the load, but warnings informing that the area was removed are **NOT** printed. And that's it ! **updaterMode** is used only for that : allowing or not allowing warnings to be printed.
As binding constraints RHS scenarii are not managed by old GUI, we don't care about this **updaterMode** for BC, and there is nothing to do here.

guilpier-code commented on src/libs/antares/study/scenario-builder/rules.cpp:
```diff
@@ -385,6 +390,14 @@ void Rules::sendWarningsForDisabledClusters()
     }
 }
 
+bool Rules::readBindingConstraints(const AreaName::Vector &splitKey, String value) {
```
Maybe move the definition of this function with the other read function in the current file.

flomnes replied:
Place after `Rules::readLink`

guilpier-code commented on src/solver/simulation/TimeSeriesNumbersWriter.h:
```diff
@@ -0,0 +1,23 @@
+#include "antares/study/binding_constraint/BindingConstraintsList.h"
+
+namespace Antares::Solver::Simulation {
+class TimeSeriesNumbersWriter: public ITimeSeriesNumbersWriter {
```
class **TimeSeriesNumbersWriter** is about binding constraint numbers. The class name does not reflect it.

Besides, I can see that this class inherits from  **ITimeSeriesNumbersWriter**, probably to use some kind of polymorphism. But I can't see any polymorphism, am I missing something ?
In this case, we should consider removing the parent class.

flomnes replied:
Dependency injection in tests

guilpier-code commented on src/tests/end-to-end/simple-study.cpp:
```diff
@@ -12,15 +23,34 @@
 using namespace Antares::Data;
 
-BOOST_AUTO_TEST_SUITE(simple_test)
+class NoOPResultWriter: public Solver::IResultWriter {
```
**NoOPResultWriter** should be renamed **nullResultWriter**, more clear.
**NoOPResultWriter** does not mean anything to me, sorry.

guilpier-code replied:
File totally changed in [PR 1457](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/1457)

guilpier-code commented on src/tests/end-to-end/simple-study.cpp:
```diff
@@ -236,15 +282,82 @@ BOOST_AUTO_TEST_CASE(one_mc_year_one_ts)
+    auto BC = addBindingConstraints(pStudy, "BC1", "Group1");
+    BC->weight(link, 1);
+    BC->enabled(true);
+    BC->mutateTypeWithoutCheck(BindingConstraint::typeHourly);
```
**Not strictly related to current PR.**
About : 
`BC->mutateTypeWithoutCheck(BindingConstraint::typeHourly);`

We should rename this function into **changeTypeTo**, more clear and shorter.
We don't need to know this call involves no check.

guilpier-code replied:
File totally changed in [PR 1457](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/1457)

guilpier-code commented on src/tests/end-to-end/simple-study.cpp:
```diff
@@ -236,15 +282,82 @@ BOOST_AUTO_TEST_CASE(one_mc_year_one_ts)
+}
+
+BOOST_AUTO_TEST_CASE(one_mc_year_one_ts__Binding_Constraints) {
+    //Create study
```
I can't see why we need so much areas and links to tests a single BC.
Maybe this test could be simplified.
For instance, we could have one area with a cluster and have a BC on that cluster.
And if we want to have a BC on a link, 2 areas and a link would suffice, right ?
Of course, these 2 cases would lie in a separate test.

guilpier-code replied:
File totally changed in [PR 1457](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/1457)

guilpier-code commented on src/tests/end-to-end/simple-study.cpp:
```diff
@@ -84,13 +114,20 @@ Area* addArea(Study::Ptr pStudy, const std::string& areaName, int nbTS)
 	return pArea;
 }
 
+std::shared_ptr<BindingConstraint> addBindingConstraints(Study::Ptr study, std::string name, std::string group) {
```
Maybe we should rename this function into **addBindingConstraintsToStudy**, more clear.
Instructions : 
```C++
bc->type();
auto ts = study->bindingConstraints.timeSeriesNumbers[group];
```
seem to be useless (at least for now).

guilpier-code replied:
Totally changed in [PR 1457](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/1457)

guilpier-code commented on src/tests/end-to-end/simple-study.cpp:
```diff
@@ -236,15 +282,82 @@ BOOST_AUTO_TEST_CASE(one_mc_year_one_ts)
+    pCluster->series->timeSeries.fillColumn(0, availablePower);
+
+    //Add BC
+    auto BC = addBindingConstraints(pStudy, "BC1", "Group1");
```
I guess that what we should have is : 
```C++
BindingConstraint BC(/*whatever args to create a decent bc*/);
BC.weight(link, 1);
BC.enabled(true);
...
addBindingConstraintsToStudy(BC, pStudy);
```

guilpier-code commented on src/tests/end-to-end/simple-study.cpp:
```diff
@@ -236,15 +282,82 @@ BOOST_AUTO_TEST_CASE(one_mc_year_one_ts)
+	cleanSimulation(pStudy, simulation);
+}
+
+BOOST_AUTO_TEST_CASE(one_mc_year_one_ts__Binding_Constraints) {
```
**one_mc_year_one_ts__Binding_Constraints** : 
The name of the test is unclear to me, which results do we expect ?

guilpier-code commented on src/tests/end-to-end/simple-study.cpp:
```diff
@@ -236,15 +282,82 @@ BOOST_AUTO_TEST_CASE(one_mc_year_one_ts)
+    auto pCluster = addCluster(area1, "Cluster 1", maximumPower, 1, nbTS);
+
+    //Initialize time series
+    pCluster->series->timeSeries.fillColumn(0, availablePower);
```
The **pCluster** (should be renamed into **cluster**) is instantiated but does not seem to be used. 
Do I miss something ?

guilpier-code commented on src/tests/end-to-end/simple-study.cpp:
```diff
@@ -236,15 +282,82 @@ BOOST_AUTO_TEST_CASE(one_mc_year_one_ts)
+
+BOOST_AUTO_TEST_CASE(one_mc_year_one_ts__Binding_Constraints) {
+    //Create study
+    Study::Ptr pStudy = std::make_shared<Study>(true); // for the solver
```
**Not strictly related to the current PR**
About **pStudy** : shouldn't it be renamed into **study**  ? 
I can't the reason there is a heading **p**, as it's not part of a fixture or class.

guilpier-code commented on src/tests/end-to-end/simple-study.cpp:
```diff
@@ -479,3 +592,292 @@ BOOST_AUTO_TEST_CASE(two_mc_year_two_ts_different_weight)
 
 BOOST_AUTO_TEST_SUITE_END()
+
+void prepareStudy(int nbYears, int nbTS, Study::Ptr &pStudy, Area *&area1,
```
ok, seemingly some tests are not finished, so my review on the current file stops here.

flomnes committed:
Don't clear Study::bindingConstraints in StudyRuntimeInfos::initializeBindingConstraints

JasonMarechal committed:
Fix warning message on conversions issue uint/size_t/singed int

JasonMarechal committed:
Reorganize end to end tests

JasonMarechal committed:
Fix windows compilation

JasonMarechal committed:
Use macro to import win32sock transparently

JasonMarechal committed:
Legacy binding constraints end at 8.7

JasonMarechal committed:
Fix not passing parameter properly for legacy studies

JasonMarechal committed:
Remove left over code

flomnes committed:
Fix opLess case when loading legacy BCs

flomnes committed:
Resize timeseries when loading legacy BCs

flomnes committed:
Use resize instead of reserve, fix size

JasonMarechal25 committed:
Test for legacy studies

JasonMarechal25 committed:
Handle legacy studies

Use "id" instead of name for time_series files to match legacy behaviour

JasonMarechal25 committed:
Code formatting

JasonMarechal25 committed:
Rename timeSeries -> RHSTimeSeries

JasonMarechal25 committed:
Disable kirchhof tests for now

flomnes committed:
[sc-builder] Don't draw BC timeseries number if only one TS (#1336)

JasonMarechal committed:
Study on stack generate overflow, declare on head instead

JasonMarechal25 committed:
Disable tests

JasonMarechal25 committed:
Update version 8.7.0-RC1

JasonMarechal25 committed:
Set Beta1

flomnes committed:
Invalidate cache by changing key, BETA++

flomnes committed:
Fix version parser

flomnes committed:
BETA++

flomnes committed:
Revert workflows

JasonMarechal committed:
Give same name and id to both constraints to ensure they are processed in the same order as before

JasonMarechal committed:
Fix unit test, less brittle to binding constraint list order

guilpier-code commented on src/libs/antares/study/action/handler/antares-study/constraint/create.cpp:
```diff
@@ -88,7 +90,7 @@ bool Create::prepareWL(Context& ctx)
     // Computing the futur ID of the area
     Antares::Data::ConstraintName id;
-    const Antares::Data::BindingConstraint* constraintFound = nullptr;
+    std::shared_ptr<Antares::Data::BindingConstraint> constraintFound = nullptr;
```
why initializing an **std::shared_ptr** to **null**, isn't that its default value anyway ?

guilpier-code replied:
Done in [PR 1466](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/1466)

guilpier-code commented on src/libs/antares/study/binding_constraint/BindingConstraint.cpp:
```diff
@@ -494,492 +334,39 @@ void BindingConstraint::copyOffsets(const Study& study,
     // The type must be `hourly` by default for studies <=3.1, which was the only
-    // type of binding constraints supported.
-    pType = typeUnknown;
+// type of binding constraints supported.
```
Comments indentation to correct

guilpier-code commented on src/libs/antares/study/binding_constraint/BindingConstraint.h:
```diff
@@ -508,199 +431,27 @@ class BindingConstraint final : public Yuni::NonCopyable<BindingConstraint>
-    template<class PredicateT>
-    void each(const PredicateT& predicate) const;
+    void copyFrom(BindingConstraint const *original);
+};;; // class BindingConstraint
```
Why so much semicolons ?

guilpier-code commented on src/libs/antares/study/binding_constraint/BindingConstraintsList.h:
```diff
@@ -0,0 +1,199 @@
+    template<class ListBindingConstraints>
+    [[nodiscard]] static unsigned int NumberOfTimeseries(const ListBindingConstraints &list, const std::string &group_name);
+
+    std::map<std::string, Data::BindingConstraintTimeSeriesNumbers, std::less<>> timeSeriesNumbers;
```
data member : **timeSeriesNumbers**
should be renamed into **groupToTSnumbers** (for example), the **std::string** for key is a group, isn't it ?

guilpier-code commented on src/solver/simulation/solver.hxx:
```diff
@@ -379,7 +380,8 @@ void ISimulation<Impl>::run()
 
         // Export ts-numbers into output
-        TimeSeriesNumbers::StoreTimeseriesIntoOuput(study);
+        BindingConstraintsTimeSeriesNumbersWriter time_series_writer(pResultWriter);
```
About : 
```C++
BindingConstraintsTimeSeriesNumbersWriter time_series_writer(pResultWriter);
TimeSeriesNumbers::StoreTimeSeriesNumbersIntoOuput(study, time_series_writer);
```
- the **time_series_writer** is an implementation detail that **solver.hxx** should not know about, it should be instantiated inside function **StoreTimeSeriesNumbersIntoOuput** we call right after.
- the **pResultWriter** we use here to instantiate **time_series_writer** is exactly the same (same address) as the one of **study** (study.resultWriter), that is of the first argument we pass to **StoreTimeSeriesNumbersIntoOuput**. This makes the second argument useless.

We should have : 
```C++
// solver.hxx
TimeSeriesNumbers::StoreTimeSeriesNumbersIntoOuput(study);
```
And : 
```C++
// timeseries-numbers.cpp
void TimeSeriesNumbers::StoreTimeSeriesNumbersIntoOuput(Data::Study& study)
{
    
    if (study.parameters.storeTimeseriesNumbers)
    {
        fixTSNumbersWhenWidthIsOne(study);
        study.storeTimeSeriesNumbers<TimeSeries::timeSeriesLoad>();
        ...
        BindingConstraintsTimeSeriesNumbersWriter writer(study.resultWriter);
        writer.write(study.bindingConstraints);
    }
}
```

guilpier-code replied:
Done in [PR #1466 ](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/1466)

guilpier-code commented on src/libs/antares/study/binding_constraint/BindingConstraintsList.hxx:
```diff
@@ -0,0 +1,75 @@
+}
+
+template<class ListBindingConstraints>
+unsigned int BindingConstraintsList::NumberOfTimeseries(const ListBindingConstraints &list, const std::string& group_name) {
```
**NumberOfTimeseries(...)** : does not need to be a template function, as it is called once

guilpier-code replied:
Currently done on **develop**

guilpier-code commented on src/libs/antares/study/binding_constraint/BindingConstraintLoader.cpp:
```diff
@@ -0,0 +1,270 @@
+    }
+
+    // Checking for validity
+    if (!bc->pName || !bc->pID || bc->pOperator == BindingConstraint::opUnknown ||
```
About the **if** statement : 
```C++
if (!bc->pName || !bc->pID || bc->pOperator == BindingConstraint::opUnknown ||
        bc->pType == BindingConstraint::typeUnknown)
```
Really seems to be useless. We should remove the **if** but leave the related block.
To make my point more clear, how should we react if we saw : 
```C++
if (!bc->pName)
{
    if (!bc->pName)
    {
          logs.error() << ...;
    }
}
```
We would think that something is redundant, right ?

guilpier-code replied:
Done in **develop**

guilpier-code commented on src/tests/end-to-end/utils/utils.cpp:
```diff
@@ -0,0 +1,98 @@
+
+using namespace Antares::Data;
+
+void prepareStudy(Study::Ptr pStudy, int nbYears)
```
Function **prepareStudy** should be named **initializeStudy**

guilpier-code replied:
This function was removed in [PR 1457](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/1457)

guilpier-code commented on src/tests/end-to-end/binding_constraints/test_binding_constraints.cpp:
```diff
@@ -0,0 +1,425 @@
+    Area *area2 = addArea(pStudy, "Area 2", nbTS);
+    area1= addArea(pStudy, "Area 1", nbTS);
+    link= AreaAddLinkBetweenAreas(area1, area2);//Prepare study
+    prepareStudy(pStudy, nbYears);
```
`prepareStudy(pStudy, nbYears); `(or **initializeStudy** as recommended in review of **utils.cpp**) should be called earlier in the test (at the first line)

guilpier-code replied:
Function removed in [PR 1457](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/1457)

guilpier-code commented on src/tests/end-to-end/binding_constraints/test_binding_constraints.cpp:
```diff
@@ -0,0 +1,425 @@
+//		area				: Area to be checked
+//		expectedHourlyValue	: Expected hourly value
+template<class VCard>
+void checkVariable(
```
As we don't use function **checkVariable** in the current file, we should remove it

guilpier-code replied:
See PR [1457](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/1457) for removal

guilpier-code commented on src/tests/end-to-end/binding_constraints/test_binding_constraints.cpp:
```diff
@@ -0,0 +1,425 @@
+    return std::pair(BC, link);
+}
+
+BOOST_AUTO_TEST_CASE(one_mc_year_one_ts__Binding_Constraints_Hourly)
```
In all these tests, we only test the flow on a link. What about testing the production of a cluster on which we would have a BC that lowers that production ?

guilpier-code replied:
See [1437](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/1437) for adding this test

JasonMarechal25 committed:
Fix typo

JasonMarechal25 committed:
Remove deadcode

JasonMarechal25 committed:
Fix indent

JasonMarechal25 committed:
Rename map TimeSeriesNumbers -> GroupToTimeSeriesNumbers

JasonMarechal25 committed:
When loading legacy study, place them in the same group

guilpier-code commented on src/libs/antares/study/binding_constraint/BindingConstraintsList.cpp:
```diff
@@ -0,0 +1,383 @@
+    return hasError;
+}
+
+void BindingConstraintsList::initializeTsNumbers() {
```
**initializeTsNumbers()** is **private** and thus cannot be called from end-to-end tests.
Should we make it **public** ? Or should it be called in another **public** function callable from tests ?

guilpier-code replied:
disappeared in current **develop**
